openmls/src/binary_tree/array_representation/tree.rs

```diff
@@ -0,0 +1,165 @@
+    /// `Result`, because it might throw a
+    /// [LibraryError](ABinaryTreeError::LibraryError).
+    pub(crate) fn leaves(&self) -> Result<Vec<&T>, ABinaryTreeError> {
+        let mut leaf_references = Vec::new();
+        for leaf_index in 0..self.leaf_count() {
+            let node_index = usize::try_from(to_node_index(leaf_index))
+                // The tree size and thus the leaf count should fit into usize on 32
+                // bit architectures.
+                .map_err(|_| ABinaryTreeError::LibraryError)?;
+            let node_ref = self
+                .nodes
+                .get(node_index)
+                // Since the index is within the bounds of the tree, this should
+                // be Some.
+                .ok_or(ABinaryTreeError::LibraryError)?;
+            leaf_references.push(node_ref);
+        }
+        Ok(leaf_references)
+    }
+
+    /// Creates and returns an empty `AbDiff`.
+    pub(crate) fn empty_diff(&self) -> AbDiff<'_, T> {
+        self.into()
+    }
+
+    /// Merges the changes applied to the `StagedAbDiff` into the tree. This
+    /// function should not fail and only returns a `Result`, because it might
+    /// throw a [LibraryError](ABinaryTreeError::LibraryError).
+    pub(crate) fn merge_diff(&mut self, diff: StagedAbDiff<T>) -> Result<(), ABinaryTreeError> {
+        // The diff size should fit into a 32 bit usize.
+        let diff_size = usize::try_from(diff.size()).map_err(|_| ABinaryTreeError::LibraryError)?;
+        // If the size of the diff is smaller than the tree, truncate the tree
+        // to the size of the diff.
+        self.nodes.truncate(diff_size);
+
+        // Iterate over the BTreeMap in order of indices.
+        for (node_index, diff_node) in diff.diff().into_iter() {
+            match node_index {
+                // If the node would extend the tree, push it to the vector of nodes.
+                node_index if node_index == self.size() => self.nodes.push(diff_node),
+                // If the node index points too far outside of the tree,
+                // something has gone wrong.
+                node_index if node_index > self.size() => {
+                    return Err(ABinaryTreeError::LibraryError)
+                }
+                // If the node_index points to somewhere within the size of the
+                // tree, do a swap-remove.
+                node_index => {
+                    // Perform swap-remove.
+                    self.nodes.push(diff_node);
+                    self.nodes.swap_remove(
+                        // If the node_index is larger that u32 max, something
+                        // has gone wrong.
+                        usize::try_from(node_index).map_err(|_| ABinaryTreeError::LibraryError)?,
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Export the nodes of the tree in the array representation.
+    pub(crate) fn export_nodes(&self) -> Vec<T> {
```

Comment:
Do we really need this? I'd prefer having a function that returns a slice and any caller can copy it if needed.